crates/arroyo-connectors/src/nats/mod.rs

```diff
@@ -0,0 +1,316 @@
+        config: OperatorConfig,
+    ) -> anyhow::Result<OperatorNode> {
+        Ok(match table.connector_type {
+            ConnectorType::Source { .. } => OperatorNode::from_source(Box::new(NatsSourceFunc {
+                stream: table.client_configs.get("nats.stream").cloned().unwrap(),
+                servers: profile.servers.clone(),
+                connection: profile.clone(),
+                table: table.clone(),
+                framing: config.framing,
+                format: config.format.unwrap(),
+                bad_data: config.bad_data,
+                consumer_config: get_nats_client_config(&profile, &table),
+                messages_per_second: NonZeroU32::new(
+                    config
+                        .rate_limit
+                        .map(|l| l.messages_per_second)
+                        .unwrap_or(u32::MAX),
+                )
+                .unwrap(),
+            })),
+            ConnectorType::Sink { .. } => OperatorNode::from_operator(Box::new(NatsSinkFunc {
+                subject: table.client_configs.get("nats.subject").cloned().unwrap(),
+                servers: profile.servers.clone(),
+                connection: profile.clone(),
+                table: table.clone(),
+                publisher: None,
+                client_config: get_nats_client_config(&profile, &table),
+                serializer: ArrowSerializer::new(config.format.unwrap()),
+            })),
+        })
+    }
+}
+
+fn get_nats_client_config(connection: &NatsConfig, table: &NatsTable) -> HashMap<String, String> {
+    let mut config: HashMap<String, String> = HashMap::new();
+    match &connection.authentication {
+        NatsConfigAuthentication::None {} => {}
+        NatsConfigAuthentication::Credentials { username, password } => {
+            config.insert(
+                "nats.username".to_string(),
+                username
+                    .sub_env_vars()
+                    .expect("Missing env-vars for NATS username"),
+            );
+            config.insert(
+                "nats.password".to_string(),
+                password
+                    .sub_env_vars()
+                    .expect("Missing env-vars for NATS password"),
+            );
+            config.extend(
+                table
+                    .client_configs
+                    .iter()
+                    .map(|(k, v)| (k.to_string(), v.to_string())),
+            );
+        }
+    };
+    config
+}
+
+async fn get_nats_client(
+    connection: &NatsConfig,
+    table: &NatsTable,
+) -> anyhow::Result<async_nats::Client> {
+    let client_config = get_nats_client_config(connection, table);
+    let servers_str = connection.servers.clone();
+    let servers_vec: Vec<ServerAddr> = servers_str.split(',').map(|s| s.parse().unwrap()).collect();
+    let client = async_nats::ConnectOptions::new()
+        .user_and_password(
+            client_config.get("nats.username").unwrap().to_string(),
```

Comment:
This currently panics if you don't specify a username and password; you might want to rework it so that the user_and_password method is only called if the username and password are Some values. For example

```rust
    let mut opts = async_nats::ConnectOptions::new();

    match (client_config.get("nats.username"), client_config.get("nats.password")) {
        (Some(username), Some(password)) => {
            opts = opts.user_and_password(username.clone(), password.clone());
        }
        _ => {}
    };
    
    
    let client = opts
        .connect(servers_vec)
        .await?;
```

Reply:
That's modified as well. Should other authentication be added to have it merged into master?